### PR TITLE
Sctp peer address params 

### PR DESCRIPTION
**Pull request by hoelscher**
Add suport for sctp_status and sctp_peer_addr_params for getsockopt

### CODE REVIEW

File: gtests/net/packetdrill/parser.y [hunk 1]
```diff
@@ -2041,7 +2047,7 @@ tcp_option
 	}
 }
 ;
-
+
```
Comment (tuexen) [inline gtests/net/packetdrill/parser.y]: Leave line empty.

File: gtests/net/packetdrill/parser.y [hunk 2]
```diff
@@ -2413,42 +2424,201 @@ sctp_sackinfo
...
+: SPP_ADDRESS '=' ELLIPSIS { $$ = new_expression(EXPR_ELLIPSIS); }
+| SPP_ADDRESS '=' sockaddr { $$ = $3; }
+;
+
+spp_hbinterval
+: SPP_HBINTERVAL '=' INTEGER { 
+	if (!is_valid_u32($3)) {
+		semantic_error("spp_hbinterval out of range");
+	}
+	$$ = new_integer_expression($3 , "%u"); }
```
Comment (tuexen) [inline gtests/net/packetdrill/parser.y]: No space between $3 and ,.

File: gtests/net/packetdrill/parser.y [hunk 3]
```diff
@@ -2413,42 +2424,201 @@ sctp_sackinfo
...
+| SPP_HBINTERVAL '=' ELLIPSIS { $$ = new_expression(EXPR_ELLIPSIS); }
+;
+
+spp_pathmtu
+: SPP_PATHMTU '=' INTEGER {
+	if (!is_valid_u32($3)) {
+		semantic_error("spp_pathmtu out of range");
+	}
+	$$ = new_integer_expression($3 , "%u"); }
```
Comment (tuexen) [inline gtests/net/packetdrill/parser.y]: No space between $3 and ,.

File: gtests/net/packetdrill/parser.y [hunk 4]
```diff
@@ -2413,42 +2424,201 @@ sctp_sackinfo
...
+| SPP_PATHMTU '=' ELLIPSIS { $$ = new_expression(EXPR_ELLIPSIS); }
+;
+
+spp_pathmaxrxt
+: SPP_PATHMAXRXT '=' INTEGER {
+	if (!is_valid_u16($3)) {
+		semantic_error("spp_pathmaxrxt out of range");
+	}
+	$$ = new_integer_expression($3 , "%u"); }
```
Comment (tuexen) [inline gtests/net/packetdrill/parser.y]: No space between $3 and ,.

File: gtests/net/packetdrill/run_system_call.c [hunk 5]
```diff
@@ -1690,43 +1728,211 @@ static int syscall_getsockopt(struct state *state, struct syscall_spec *syscall,
...
+		struct sctp_status *live_status = live_optval;
+		int state, rwnd, fragmentation_point;
+		short unackdata, penddata, instrms, outstrms;
+		if (sstat_state->type != EXPR_ELLIPSIS) {
+			if (get_s32(sstat_state, &state, error)) {
+				free(live_optval);
+				return STATUS_ERR;
+			}
+			if (live_status->sstat_state != state) {
+				asprintf(error, "Bad getsockopt SCTP_STATUS state: expected: %d actual: %d ",
```
Comment (tuexen) [inline gtests/net/packetdrill/run_system_call.c]: Remove final space

File: gtests/net/packetdrill/run_system_call.c [hunk 6]
```diff
@@ -1690,43 +1728,211 @@ static int syscall_getsockopt(struct state *state, struct syscall_spec *syscall,
...
+				return STATUS_ERR;
+			}
+		}
+		if (sstat_rwnd->type != EXPR_ELLIPSIS) {
+			if (get_s32(sstat_rwnd, &rwnd, error)) {
+				free(live_optval);
+				return STATUS_ERR;
+			}
+			if (live_status->sstat_rwnd != rwnd) {
+				asprintf(error, "Bad getsockopt SCTP_STATUS rwnd: expected: %u actual: %u ",
```
Comment (tuexen) [inline gtests/net/packetdrill/run_system_call.c]: Remove final space

File: gtests/net/packetdrill/run_system_call.c [hunk 7]
```diff
@@ -1690,43 +1728,211 @@ static int syscall_getsockopt(struct state *state, struct syscall_spec *syscall,
...
+		struct sctp_paddrparams *live_params = live_optval;
+		int hbinterval, pathmtu;
+		short pathmaxrxt;
+		if (spp_hbinterval->type != EXPR_ELLIPSIS) {
+			if (get_s32(spp_hbinterval, &hbinterval, error)) {
+				free(live_optval);
+				return STATUS_ERR;
+			}
+			if (live_params->spp_hbinterval != hbinterval) {
+				asprintf(error, "Bad getsockopt SCTP_PARAMS hbinterval: expected: %d actual: %d ",
```
Comment (tuexen) [inline gtests/net/packetdrill/run_system_call.c]: Isn't hbinterval unsigned?

File: gtests/net/packetdrill/run_system_call.c [hunk 8]
```diff
@@ -1690,43 +1728,211 @@ static int syscall_getsockopt(struct state *state, struct syscall_spec *syscall,
...
+				return STATUS_ERR;
+			}
+		}
+		if (spp_pathmaxrxt->type != EXPR_ELLIPSIS) {
+			if (get_s16(spp_pathmaxrxt, &pathmaxrxt, error)) {
+				free(live_optval);
+				return STATUS_ERR;
+			}
+			if (live_params->spp_pathmaxrxt != pathmaxrxt) {
+				asprintf(error, "Bad getsockopt SCTP_PARAMS pathmaxrxt: expected: %d actual: %d ",
```
Comment (tuexen) [inline gtests/net/packetdrill/run_system_call.c]: Isn't pathmaxrxt unsigned?

File: gtests/net/packetdrill/run_system_call.c [hunk 9]
```diff
@@ -1690,43 +1728,211 @@ static int syscall_getsockopt(struct state *state, struct syscall_spec *syscall,
...
+				return STATUS_ERR;
+			}
+		}
+		if (spp_pathmtu->type != EXPR_ELLIPSIS) {
+			if (get_s32(spp_pathmtu, &pathmtu, error)) {
+				free(live_optval);
+				return STATUS_ERR;
+			}
+			if (live_params->spp_pathmtu != pathmtu) {
+				asprintf(error, "Bad getsockopt SCTP_PARAMS pathmtu: expected: %d actual: %d ",
```
Comment (tuexen) [inline gtests/net/packetdrill/run_system_call.c]: Isn't pathmtu unsigned?

File: gtests/net/packetdrill/symbols_freebsd.c [hunk 10]
```diff
@@ -96,6 +96,7 @@ struct int_symbol platform_symbols_table[] = {
 	{ SCTP_SHUTDOWN_RECEIVED,           "SCTP_SHUTDOWN_RECEIVED"          },
 	{ SCTP_SHUTDOWN_ACK_SENT,           "SCTP_SHUTDOWN_ACK_SENT"          },
 	{ SCTP_SHUTDOWN_PENDING,            "SCTP_SHUTDOWN_PENDING"           },
+	{ SCTP_PEER_ADDR_PARAMS,            "SCTP_PEER_ADDR_PARAMS"           },
```
Comment (tuexen) [inline gtests/net/packetdrill/symbols_freebsd.c]: Add this line after SCTP_MAX_BURST. This keeps socket options together.

File: gtests/net/packetdrill/symbols_linux.c [hunk 11]
```diff
@@ -126,6 +126,9 @@ struct int_symbol platform_symbols_table[] = {
 	{ SCTP_BOUND,                       "SCTP_BOUND"                      },
 	{ SCTP_LISTEN,                      "SCTP_LISTEN"                     },
 #endif
+#ifdef SCTP_PEER_ADDR_PARAMS
+	{ SCTP_PEER_ADDR_PARAMS,            "SCTP_PEER_ADDR_PARAMS"           },	
+#endif
```
Comment (tuexen) [inline gtests/net/packetdrill/symbols_linux.c]: Same as above.

File: gtests/net/packetdrill/tests/bsd/sctp/sctp_get_socket_options.pkt [hunk 12]
```diff
@@ -10,8 +10,34 @@
...
++0 getsockopt(3, IPPROTO_SCTP, SCTP_STATUS, {sstat_state=8, sstat_rwnd=1500, sstat_unackdata=0, sstat_penddata=0, sstat_instrms=..., 
+	sstat_outstrms=1, sstat_fragmentation_point=1452, sstat_primary=... }, [176])= 0
++0 getsockopt(3, IPPROTO_SCTP, SCTP_STATUS, {sstat_state=8, sstat_rwnd=1500, sstat_unackdata=0, sstat_penddata=0, sstat_instrms=1, 
+	sstat_outstrms=..., sstat_fragmentation_point=1452, sstat_primary=... }, [176])= 0
++0 getsockopt(3, IPPROTO_SCTP, SCTP_STATUS, {sstat_state=8, sstat_rwnd=1500, sstat_unackdata=0, sstat_penddata=0, sstat_instrms=1, 
+	sstat_outstrms=1, sstat_fragmentation_point=..., sstat_primary=... }, [176])= 0
++0 getsockopt(3, IPPROTO_SCTP, SCTP_STATUS, {sstat_state=8, sstat_rwnd=1500, sstat_unackdata=0, sstat_penddata=0, sstat_instrms=1, 
+	sstat_outstrms=1, sstat_fragmentation_point=..., 
+	sstat_primary={ spinfo_state=1, spinfo_cwnd=4464, spinfo_srtt=..., spinfo_rto=1000, spinfo_mtu=1468} }, [176])= 0
+
```
Comment (tuexen) [inline gtests/net/packetdrill/tests/bsd/sctp/sctp_get_socket_options.pkt]: Use symbolic names for sstat_state and spinfo_state.

File: gtests/net/packetdrill/run_system_call.c (hunk 7 above)
Comment (hoelscher) [inline gtests/net/packetdrill/run_system_call.c]: from https://tools.ietf.org/html/draft-ietf-tsvwg-sctpsocket-13 and from bsd:
struct sctp_paddrparams {
        struct sockaddr_storage spp_address;
        sctp_assoc_t spp_assoc_id;
        uint32_t spp_hbinterval;
        uint32_t spp_pathmtu;
        uint32_t spp_flags;
        uint32_t spp_ipv6_flowlabel;
        uint16_t spp_pathmaxrxt;
        uint8_t spp_dscp;
};

yes it is unsigned

**Pull request merged**